odl/trafos/util/ft_utils.py

```diff
@@ -0,0 +1,654 @@
+    number of points, while for an even size, the second option
+    guarantees that 0 is contained.
+
+    If a real-to-complex (half-complex) transform is to be computed,
+    the reciprocal grid has the shape ``M[i] = floor(N[i]/2) + 1``
+    in the last transform axis ``i``.
+
+    Parameters
+    ----------
+    grid : `RegularGrid`
+        Original sampling grid
+    shift : bool or `sequence` of bools, optional
+        If ``True``, the grid is shifted by half a stride in the negative
+        direction. With a sequence, this option is applied separately on
+        each axis.
+    axes : int or `sequence` of ints, optional
+        Dimensions in which to calculate the reciprocal. The sequence
+        must have the same length as ``shift`` if the latter is given
+        as a sequence. ``None`` means all axes in ``grid``.
+    halfcomplex : bool, optional
+        If ``True``, return the half of the grid with last coordinate
+        less than zero. This is related to the fact that for real-valued
+        functions, the other half is the mirrored complex conjugate of
+        the given half and therefore needs not be stored.
+
+    Returns
+    -------
+    recip : `RegularGrid`
+        The reciprocal grid.
+    """
+    if axes is None:
+        axes = list(range(grid.ndim))
+    else:
+        try:
+            axes = [int(axes)]
+        except TypeError:
+            axes = list(axes)
+
+    # List indicating shift or not per "active" axis, same length as axes
+    shift_list = normalized_scalar_param_list(shift, length=len(axes),
+                                              param_conv=bool)
+
+    # Full-length vectors
+    stride = grid.stride
+    shape = np.array(grid.shape)
+    rmin = grid.min_pt.copy()
+    rmax = grid.max_pt.copy()
+    rshape = list(shape)
+
+    # Shifted axes (full length to avoid ugly double indexing)
+    shifted = np.zeros(grid.ndim, dtype=bool)
+    shifted[axes] = shift_list
+    rmin[shifted] = -np.pi / stride[shifted]
+    # Length min->max increases by double the shift, so we
+    # have to compensate by a full stride
+    rmax[shifted] = (-rmin[shifted] -
+                     2 * np.pi / (stride[shifted] * shape[shifted]))
+
+    # Non-shifted axes
+    not_shifted = np.zeros(grid.ndim, dtype=bool)
+    not_shifted[axes] = np.logical_not(shift_list)
+    rmin[not_shifted] = ((-1.0 + 1.0 / shape[not_shifted]) *
+                         np.pi / stride[not_shifted])
+    rmax[not_shifted] = -rmin[not_shifted]
+
+    # Change last axis shape and max if halfcomplex
+    if halfcomplex:
+        rshape[axes[-1]] = shape[axes[-1]] // 2 + 1
+
+        # - Odd and not shifted or even and shifted -> 0
+        # - Odd and shifted -> - stride / 2
+        # - Even and not shifted -> + stride / 2
+        last_odd = shape[axes[-1]] % 2 == 1
+        last_shifted = shift_list[-1]
+        half_rstride = np.pi / (shape[axes[-1]] * stride[axes[-1]])
+
+        if last_odd:
```

Comment:
The suggestion is not quite right, it misses the sign difference between the two cases. Need 3 clauses.